src/components/ble/AppleNotificationCenterClient.cpp

```diff
@@ -0,0 +1,514 @@
+        NRF_LOG_INFO(msg);
+        dataSourceHandle = characteristic->val_handle;
+        isDataCharacteristicDiscovered = true;
+      }
+    }
+  }
+  return 0;
+}
+
+int AppleNotificationCenterClient::OnDescriptorDiscoveryEventCallback(uint16_t connectionHandle,
+                                                                      const ble_gatt_error* error,
+                                                                      uint16_t characteristicValueHandle,
+                                                                      const ble_gatt_dsc* descriptor) {
+  if (error->status == 0) {
+    if (characteristicValueHandle == notificationSourceHandle && ble_uuid_cmp(&notificationSourceChar.u, &descriptor->uuid.u)) {
+      if (notificationSourceDescriptorHandle == 0) {
+        NRF_LOG_INFO("ANCS Descriptor discovered : %d", descriptor->handle);
+        notificationSourceDescriptorHandle = descriptor->handle;
+        isDescriptorFound = true;
+        uint8_t value[2] {1, 0};
+        ble_gattc_write_flat(connectionHandle, notificationSourceDescriptorHandle, value, sizeof(value), NewAlertSubcribeCallback, this);
+        ble_gattc_write_flat(connectionHandle, ancsEndHandle, value, sizeof(value), NewAlertSubcribeCallback, this);
+      }
+    } else if (characteristicValueHandle == controlPointHandle && ble_uuid_cmp(&controlPointChar.u, &descriptor->uuid.u)) {
+      if (controlPointDescriptorHandle == 0) {
+        NRF_LOG_INFO("ANCS Descriptor discovered : %d", descriptor->handle);
+        controlPointDescriptorHandle = descriptor->handle;
+        isControlDescriptorFound = true;
+      }
+    } else if (characteristicValueHandle == dataSourceHandle && ble_uuid_cmp(&dataSourceChar.u, &descriptor->uuid.u)) {
+      if (dataSourceDescriptorHandle == 0) {
+        NRF_LOG_INFO("ANCS Descriptor discovered : %d", descriptor->handle);
+        dataSourceDescriptorHandle = descriptor->handle;
+        isDataDescriptorFound = true;
+        uint8_t value[2] {1, 0};
+        ble_gattc_write_flat(connectionHandle, dataSourceDescriptorHandle, value, sizeof(value), NewAlertSubcribeCallback, this);
+      }
+    }
+  } else {
+    if (error->status != BLE_HS_EDONE) {
+      char errorStr[55];
+      snprintf(errorStr, sizeof(errorStr), "ANCS Descriptor discovery ERROR: %d", error->status);
+      NRF_LOG_INFO(errorStr);
+    }
+    if (isDescriptorFound == isDataDescriptorFound)
+      onServiceDiscovered(connectionHandle);
+  }
+  return 0;
+}
+
+int AppleNotificationCenterClient::OnNewAlertSubcribe(uint16_t connectionHandle,
+                                                      const ble_gatt_error* error,
+                                                      ble_gatt_attr* /*attribute*/) {
+  if (error->status == 0) {
+    NRF_LOG_INFO("ANCS New alert subscribe OK");
+  } else {
+    NRF_LOG_INFO("ANCS New alert subscribe ERROR");
+  }
+  if (isDescriptorFound == isControlDescriptorFound && isDescriptorFound == isDataDescriptorFound)
```

Comment:
This is hard to read. What does it mean semantically to compare the values of `isDescriptorFound` and `isControlDescriptorFound` (for example)? Is the meaning the same if they're both `false` compared to when they're both `true`?